allennlp/modules/elmo.py

```diff
@@ -214,7 +217,7 @@ def from_params(cls, params: Params) -> 'Elmo':
                    dropout=dropout)
 
 
-def batch_to_ids(batch: List[List[str]]) -> torch.Tensor:
+def batch_to_ids(batch: List[List[str]], vocab: Vocabulary = None) -> Tuple[torch.Tensor, torch.Tensor]:
```

Comment:
I think we still need the word `vocab` here to return the correct word ids.

If we replace with `vocab = Vocabulary()` (like it's done currently), the word ids return `1` for all words.

Reply:
Right, but my point is that you don't actually need to pass word ids around at all - the only place you make use of them is [here](https://github.com/allenai/allennlp/pull/1725/files#diff-941a4360ef08efef82e2b72210a7ab93R756), but all you are using them for is to get an empty tensor on the correct device. The approach of splitting up the softmax is nice, but you don't need to pass word ids to do that, unless I am mis-understanding something. Instead, you can just iterate over a fixed chunk size, computing the softmax in parts as you are doing currently.

Reply:
We still need the `word_ids` here (see `targets_flat`): https://github.com/allenai/allennlp/pull/1725/files#diff-941a4360ef08efef82e2b72210a7ab93R791

So `_chunked_log_probs` does not return the probabilities of the whole vocab, but just the probabilities of the target `word_ids`.

Reply:
Oh! Sorry that wasn't clear from the [documentation](https://github.com/allenai/allennlp/pull/1725/files#diff-941a4360ef08efef82e2b72210a7ab93R771) of the class. Could you remove this functionality? I think this is something which is easier to do as a post-processing step and it would considerably simplify the code. It's also not something I anticipate a lot of people wanting to do.

Reply:
Got it, so should I keep the changes made in the PR in `commands/elmo.py`? I don't think anyone would want the full softmax written to a file.

* [ ] remove Vocab dependency and output the full softmax
* [ ] refine documentation of https://github.com/allenai/allennlp/pull/1725/files#diff-941a4360ef08efef82e2b72210a7ab93R771

Reply:
Hi, so I've mostly addressed your comments. Removed vocabulary dependency from the `elmo` command and also `batch_to_id`. I still have the `softmax_vocab_file` argument (now optional) in `_ElmoSoftmax` if people wanted to do some post-processing.

You can test here:

```
> cat sample_sents.txt
hello
how are you

> allennlp elmo sample_sents.txt out2.hdf5 --top \
        --softmax-weight-file https://s3-us-west-2.amazonaws.com/allennlp/models/elmo/2x4096_512_2048cnn_2xhighway/elmo_2x4096_512_2048cnn_2xhighway_softmax_weights.hdf5
```

And then you can see the output:

```
> python -c "import h5py; f = h5py.File('out2.hdf5'); print(f['0'][:], f['0'].shape)"
```

Since the entire softmax is outputted, this file will be huge 😅. I think it would be better to output the probability of the original word which would require the vocab (so you can compute perplexities, play with language model outputs, probabilistic data), or I could simply remove it from the `elmo` command.